File: apps/cli/clients/typescript/bin/commands/revert.ts
import * as path from "node:path";
import { isCancel, select } from "@clack/prompts";
import axios from "axios";
import chalk from "chalk";
import type { Command } from "commander";
import fs from "fs-extra";
import { getDokugenBackupPath } from "../helpers/fileOps.js";
import { getUserInfo } from "../helpers/git.js";
import { checkAndUpdate, getBackendDomain } from "../helpers/network.js";

export function registerRevertCommand(program: Command) {
  const projectName = path.basename(process.cwd());

  program
    .command("revert")
    .description(`Revert ${projectName} README.md to the previous Dokugen-generated backup`)
    .action(async () => {
      await checkAndUpdate();

      const projectDir = process.cwd();
      const backupFile = getDokugenBackupPath(projectDir);
      const readmePath = path.join(projectDir, "README.md");

      if (!(await fs.pathExists(backupFile))) {
        console.log(chalk.red("No backup found. Run 'dokugen generate' or 'dokugen update' first to create one."));
        process.exit(1);
      }

      console.log(chalk.blue("Found a previous README backup."));

      const action = await select({
        message: `Revert ${projectName} README.md to the previous Dokugen-generated version?`,
        options: [
          { value: "yes", label: "Yes, revert it" },
          { value: "no", label: "No, keep current" },
        ],
      });

      if (isCancel(action) || action === "no") {
        console.log(chalk.yellow("Revert cancelled."));
        return;
      }

      try {
        const backupContent = await fs.readFile(backupFile, "utf-8");
        await fs.writeFile(readmePath, backupContent, "utf-8");
        console.log(chalk.green("README.md successfully reverted to the previous version!"));
        // Fire-and-forget usage tracking
        try {
          const backendDomain = await getBackendDomain();
          const userInfo = getUserInfo();
          if (userInfo?.username && userInfo?.email) {
            axios.post(`${backendDomain}/api/track`, { userInfo, usageType: "revert" }).catch(() => {});
          }
        } catch {
          /* never block the user */
        }
      } catch (error: unknown) {
        const err = error as Error;
        console.error(chalk.red("Failed to revert README:"), err.message);
        process.exit(1);
      }
    });
}

File: apps/cli/clients/python/src/dokugen/commands/revert.py
import os
import sys
import threading
import requests
import questionary
from rich.console import Console
from dokugen import utils

console = Console()


def ask_yes_no(message):
    try:
        answer = questionary.select(
            message,
            choices=["Yes", "No"],
        ).ask()
        if answer is None:
            return "cancel"
        return "yes" if answer == "Yes" else "no"
    except KeyboardInterrupt:
        return "cancel"


def cmd_revert(args):
    utils.check_and_update()
    project_dir = os.getcwd()
    backup_file = utils.get_dokugen_backup_path(project_dir)

    if not os.path.exists(backup_file):
        console.print("[red]No backup found. Run 'dokugen generate' or 'dokugen update' first to create one.[/red]")
        sys.exit(1)

    console.print("[blue]Found a previous README backup.[/blue]")
    project_name = os.path.basename(project_dir)
    ans = ask_yes_no(f"Revert {project_name} README.md to the previous Dokugen-generated version?")
    if ans != "yes":
        console.print("[yellow]Revert cancelled.[/yellow]")
        return

    _, err = utils.revert_readme_from_disk(project_dir)
    if err:
        console.print(f"[red]Failed to revert README: {err}[/red]")
        sys.exit(1)

    console.print("[green]README.md successfully reverted to the previous version![/green]")

    # Fire-and-forget usage tracking
    try:
        backend_domain = utils.get_backend_domain()
        user_info = utils.get_user_info()
        if user_info and user_info.get("username") and user_info.get("email"):
            threading.Thread(
                target=lambda: requests.post(
                    f"{backend_domain}/api/track",
                    json={"userInfo": user_info, "usageType": "revert"},
                    timeout=5
                ),
                daemon=True
            ).start()
    except Exception:
        pass


def register_revert_parser(subparsers):
    project_name = os.path.basename(os.getcwd())
    subparsers.add_parser("revert", help=f"Revert {project_name} README.md to the previous Dokugen-generated backup")
